homeassistant/components/esphome/voice_assistant.py

```diff
@@ -177,19 +186,119 @@ def _event_callback(self, event: PipelineEvent) -> None:
                 "code": event.data["code"],
                 "message": event.data["message"],
             }
-            self.handle_finished()
+            self._tts_done.set()
+            error = True
 
         self.handle_event(event_type, data_to_send)
+        if error:
+            self.handle_finished()
+
+    async def _wait_for_speech(
+        self,
+        segmenter: VoiceCommandSegmenter,
+        chunk_buffer: MutableSequence[bytes],
+    ) -> bool:
+        """Buffer audio chunks until speech is detected.
+
+        Returns True if speech was detected, False otherwise.
+        """
+        # Timeout if no audio comes in for a while.
+        async with async_timeout.timeout(self.audio_timeout):
+            chunk = await self.queue.get()
+
+        while chunk:
+            segmenter.process(chunk)
+            if segmenter.in_command:
+                return True
+
+            # Buffer until command starts
+            chunk_buffer.append(chunk)
+
+            async with async_timeout.timeout(self.audio_timeout):
+                chunk = await self.queue.get()
+
+        return False
+
+    async def _segment_audio(
+        self,
+        segmenter: VoiceCommandSegmenter,
+        chunk_buffer: Sequence[bytes],
+    ) -> AsyncIterable[bytes]:
+        """Yield audio chunks until voice command has finished."""
+        # Buffered chunks first
+        for buffered_chunk in chunk_buffer:
+            yield buffered_chunk
+
+        # Timeout if no audio comes in for a while.
+        # This means the caller hung up.
+        async with async_timeout.timeout(self.audio_timeout):
+            chunk = await self.queue.get()
+
+        while chunk:
+            if not segmenter.process(chunk):
+                # Voice command is finished
+                break
+
+            yield chunk
+
+            async with async_timeout.timeout(self.audio_timeout):
+                chunk = await self.queue.get()
 
     async def run_pipeline(
         self,
+        conversation_id: str | None,
         pipeline_timeout: float = 30.0,
     ) -> None:
         """Run the Voice Assistant pipeline."""
         try:
             tts_audio_output = (
                 "raw" if self.device_info.voice_assistant_version >= 2 else "mp3"
             )
+
+            if self.device_info.voice_assistant_version >= 3:
```

Comment:
All this code can be moved outside of the `try…except`. Maybe even broken out in a method for readability ?